edb/pgsql/resolver/command.py

```diff
@@ -19,75 +19,58 @@
 """SQL resolver that compiles public SQL to internal SQL which is executable
 in our internal Postgres instance."""
 
-from typing import Optional, Dict, List
+from typing import Optional, Dict
 
 from edb.pgsql import ast as pgast
 
 from . import dispatch
 from . import context
+from . import expr as pg_res_expr
 
 Context = context.ResolverContextLevel
 
 
 @dispatch._resolve.register
 def resolve_CopyStmt(stmt: pgast.CopyStmt, *, ctx: Context) -> pgast.CopyStmt:
 
-    # Query
-    query = dispatch.resolve_opt(stmt.query, ctx=ctx)
-    relation: Optional[pgast.Relation] = None
-    col_names: Optional[List[str]] = None
+    query: Optional[pgast.Query]
 
-    if stmt.relation:
-        # A table is going to be copied, which potentially is a view that
-        # cannot be copied as a table, but needs to be wrapped into a
-        # `SELECT * FROM view`.
+    if stmt.query:
+        query = dispatch.resolve(stmt.query, ctx=ctx)
+
+    elif stmt.relation:
         relation, table = dispatch.resolve_relation(stmt.relation, ctx=ctx)
-        col_map: Dict[str, str] = {
-            col.name: col.kind.reference_as
-            for col in table.columns
-            if col.name and isinstance(col.kind, context.ColumnByName)
-        }
         if stmt.colnames:
-            col_names = [col_map[name] for name in stmt.colnames]
+            col_map: Dict[str, context.Column] = {
+                col.name: col for col in table.columns
+            }
+            selected_columns = (col_map[name] for name in stmt.colnames)
+        else:
+            selected_columns = (c for c in table.columns if not c.hidden)
+
+        # The relation being copied is potentially a view and views cannot be
+        # copied if referenced by name, so we just always wrap it into a SELECT.
 
-        if relation.schemaname == 'edgedbpub':
```

Comment:
Special case wasn't needed?

Reply:
It doesn't hurt to always wrap the relation into a SELECT. This if assumes that we only expose VIEWs as TABLEs in `edgedbpub` schema, which is probably true, but also something can and will change with the trampolined schemas.